File: netlify/functions/shared/ingredients.mjs
/** Shopping list: unit conversion, categories (aligned with Python utils/ingredients.py). */

const UNIT_CONVERSIONS = {
  ml: 1,
  millilitre: 1,
  millilitres: 1,
  l: 1000,
  litre: 1000,
  litres: 1000,
  tasse: 250,
  tasses: 250,
  "c. à soupe": 15,
  "c. à thé": 5,
  "cuillère à soupe": 15,
  "cuillères à soupe": 15,
  "cuillère à thé": 5,
  "cuillères à thé": 5,
  g: 1,
  gramme: 1,
  grammes: 1,
  kg: 1000,
  kilogramme: 1000,
  kilogrammes: 1000,
  oz: 28.35,
  once: 28.35,
  onces: 28.35,
  lb: 453.59,
  livre: 453.59,
  livres: 453.59,
};

const VOLUME_UNITS = new Set([
  "ml",
  "millilitre",
  "millilitres",
  "l",
  "litre",
  "litres",
  "tasse",
  "tasses",
  "c. à soupe",
  "c. à thé",
  "cuillère à soupe",
  "cuillères à soupe",
  "cuillère à thé",
  "cuillères à thé",
]);

const WEIGHT_UNITS = new Set([
  "g",
  "gramme",
  "grammes",
  "kg",
  "kilogramme",
  "kilogrammes",
  "oz",
  "once",
  "onces",
  "lb",
  "livre",
  "livres",
]);

const INGREDIENT_CATEGORIES = {
  "Produits laitiers": ["lait", "beurre", "crème", "fromage", "yogourt", "yaourt"],
  "Viandes et poissons": ["poulet", "boeuf", "porc", "poisson", "saumon", "thon", "bacon", "jambon"],
  "Fruits et légumes": ["tomate", "oignon", "ail", "carotte", "pomme", "banane", "laitue", "épinard", "poivron"],
  "Céréales et pains": ["pain", "farine", "riz", "pâtes", "céréales", "avoine"],
  "Condiments et épices": ["sel", "poivre", "huile", "vinaigre", "sauce", "épice", "herbe"],
  Conserves: ["conserve", "boîte"],
  Surgelés: ["surgelé", "congelé"],
};

const COMMON_PANTRY_ITEMS = ["sel", "poivre", "huile", "farine", "sucre", "eau"];

function normalizeUnit(unit) {
  const unitLower = (unit || "").toLowerCase().trim();
  if (!unitLower || !(unitLower in UNIT_CONVERSIONS)) {
    return [unitLower, 1, "other"];
  }
  const factor = UNIT_CONVERSIONS[unitLower];
  if (VOLUME_UNITS.has(unitLower)) return ["ml", factor, "volume"];
  if (WEIGHT_UNITS.has(unitLower)) return ["g", factor, "weight"];
  return [unitLower, 1, "other"];
}

function convertQuantity(quantity, fromUnit) {
  const [baseUnit, factor, unitType] = normalizeUnit(fromUnit);
  if (unitType === "other") {
    return [quantity, fromUnit];
  }
  const normalizedQuantity = quantity * factor;
  if (unitType === "volume") {
    if (normalizedQuantity >= 1000) return [normalizedQuantity / 1000, "l"];
    return [normalizedQuantity, "ml"];
  }
  if (unitType === "weight") {
    if (normalizedQuantity >= 1000) return [normalizedQuantity / 1000, "kg"];
    return [normalizedQuantity, "g"];
  }
  return [normalizedQuantity, baseUnit];
}

function categorizeIngredient(ingredientName) {
  const nameLower = ingredientName.toLowerCase();
  for (const [category, keywords] of Object.entries(INGREDIENT_CATEGORIES)) {
    for (const keyword of keywords) {
      if (nameLower.includes(keyword)) return category;
    }
  }
  return "Autres";
}

function isPantryItem(ingredientName) {
  const nameLower = ingredientName.toLowerCase();
  return COMMON_PANTRY_ITEMS.some((item) => nameLower.includes(item));
}

function processIngredient(ingredient, aggregated, excludePantry) {
  const name = (ingredient.nom || "").trim();
  const quantity = Number(
    ingredient.quantité ?? ingredient.quantite ?? 0
  );
  const unit = ingredient.unité ?? ingredient.unite ?? "";

  if (!name) return;
  if (excludePantry && isPantryItem(name)) return;

  const [convertedQty, convertedUnit] = convertQuantity(quantity, unit || "");
  const category = categorizeIngredient(name);
  const key = `${name.toLowerCase()}\t${String(convertedUnit).toLowerCase()}`;

  if (aggregated[key]) {
    aggregated[key].quantite += convertedQty;
  } else {
    aggregated[key] = {
      nom: name,
      quantite: convertedQty,
      unite: convertedUnit,
      category,
    };
  }
}

export function aggregateIngredients(recipes, excludePantry = false) {
  const aggregated = {};

  for (const recipe of recipes) {
    const ingredients = recipe.ingredients || [];
    for (const ing of ingredients) {
      if (ing.groupe) {
        for (const item of ing.liste || []) {
          processIngredient(item, aggregated, excludePantry);
        }
      } else {
        processIngredient(ing, aggregated, excludePantry);
      }
    }
  }

  const categorized = {};
  for (const data of Object.values(aggregated)) {
    const cat = data.category;
    if (!categorized[cat]) categorized[cat] = [];
    categorized[cat].push({
      nom: data.nom,
      quantite: Math.round(data.quantite * 100) / 100,
      unite: data.unite,
    });
  }
  return categorized;
}

export function roundToPractical(quantity, unit) {
  const u = String(unit || "").toLowerCase();
  if (
    ["unité", "unités", "", "pièce", "pièces"].includes(u) ||
    u === ""
  ) {
    return Math.round(quantity + 0.49);
  }
  if (quantity < 10) return Math.round(quantity * 10) / 10;
  if (quantity >= 100) return Math.round(quantity / 10) * 10;
  return Math.round(quantity);
}

File: netlify/functions-python/utils/ingredients.py
"""Utilities for ingredient processing and shopping list generation"""

UNIT_CONVERSIONS = {
    # Volume conversions (all to ml)
    'ml': 1,
    'millilitre': 1,
    'millilitres': 1,
    'l': 1000,
    'litre': 1000,
    'litres': 1000,
    'tasse': 250,
    'tasses': 250,
    'c. à soupe': 15,
    'c. à thé': 5,
    'cuillère à soupe': 15,
    'cuillères à soupe': 15,
    'cuillère à thé': 5,
    'cuillères à thé': 5,

    # Weight conversions (all to grams)
    'g': 1,
    'gramme': 1,
    'grammes': 1,
    'kg': 1000,
    'kilogramme': 1000,
    'kilogrammes': 1000,
    'oz': 28.35,
    'once': 28.35,
    'onces': 28.35,
    'lb': 453.59,
    'livre': 453.59,
    'livres': 453.59,
}

INGREDIENT_CATEGORIES = {
    'Produits laitiers': ['lait', 'beurre', 'crème', 'fromage', 'yogourt', 'yaourt'],
    'Viandes et poissons': ['poulet', 'boeuf', 'porc', 'poisson', 'saumon', 'thon', 'bacon', 'jambon'],
    'Fruits et légumes': ['tomate', 'oignon', 'ail', 'carotte', 'pomme', 'banane', 'laitue', 'épinard', 'poivron'],
    'Céréales et pains': ['pain', 'farine', 'riz', 'pâtes', 'céréales', 'avoine'],
    'Condiments et épices': ['sel', 'poivre', 'huile', 'vinaigre', 'sauce', 'épice', 'herbe'],
    'Conserves': ['conserve', 'boîte'],
    'Surgelés': ['surgelé', 'congelé'],
}

COMMON_PANTRY_ITEMS = [
    'sel', 'poivre', 'huile', 'farine', 'sucre', 'eau'
]

def normalize_unit(unit: str) -> tuple:
    """
    Normalize unit to base unit and return conversion factor
    Returns: (base_unit, factor, unit_type)
    """
    unit_lower = unit.lower().strip()

    # Check volume units
    if unit_lower in UNIT_CONVERSIONS:
        factor = UNIT_CONVERSIONS[unit_lower]
        if unit_lower in ['ml', 'millilitre', 'millilitres', 'l', 'litre', 'litres', 'tasse', 'tasses', 'c. à soupe', 'c. à thé', 'cuillère à soupe', 'cuillères à soupe', 'cuillère à thé', 'cuillères à thé']:
            return ('ml', factor, 'volume')

    # Check weight units
    if unit_lower in UNIT_CONVERSIONS:
        factor = UNIT_CONVERSIONS[unit_lower]
        if unit_lower in ['g', 'gramme', 'grammes', 'kg', 'kilogramme', 'kilogrammes', 'oz', 'once', 'onces', 'lb', 'livre', 'livres']:
            return ('g', factor, 'weight')

    # Return original unit if no conversion available
    return (unit_lower, 1, 'other')

def convert_quantity(quantity: float, from_unit: str, to_unit: str = None) -> tuple:
    """
    Convert quantity to normalized unit
    Returns: (converted_quantity, base_unit)
    """
    base_unit, factor, unit_type = normalize_unit(from_unit)

    if unit_type == 'other':
        return (quantity, from_unit)

    normalized_quantity = quantity * factor

    # Convert to more readable units if large quantities
    if unit_type == 'volume':
        if normalized_quantity >= 1000:
            return (normalized_quantity / 1000, 'l')
        return (normalized_quantity, 'ml')

    if unit_type == 'weight':
        if normalized_quantity >= 1000:
            return (normalized_quantity / 1000, 'kg')
        return (normalized_quantity, 'g')

    return (normalized_quantity, base_unit)

def categorize_ingredient(ingredient_name: str) -> str:
    """Categorize ingredient based on its name"""
    name_lower = ingredient_name.lower()

    for category, keywords in INGREDIENT_CATEGORIES.items():
        for keyword in keywords:
            if keyword in name_lower:
                return category

    return 'Autres'

def is_pantry_item(ingredient_name: str) -> bool:
    """Check if ingredient is a common pantry item"""
    name_lower = ingredient_name.lower()
    return any(item in name_lower for item in COMMON_PANTRY_ITEMS)

def aggregate_ingredients(recipes: list, exclude_pantry: bool = False) -> dict:
    """
    Aggregate ingredients from multiple recipes
    Returns dict with categories as keys
    """
    aggregated = {}

    for recipe in recipes:
        ingredients = recipe.get('ingredients', [])

        for ing in ingredients:
            # Handle grouped ingredients
            if 'groupe' in ing:
                for item in ing.get('liste', []):
                    process_ingredient(item, aggregated, exclude_pantry)
            else:
                process_ingredient(ing, aggregated, exclude_pantry)

    # Organize by category
    categorized = {}
    for key, data in aggregated.items():
        category = data['category']
        if category not in categorized:
            categorized[category] = []

        categorized[category].append({
            'nom': data['nom'],
            'quantite': round(data['quantite'], 2),
            'unite': data['unite']
        })

    return categorized

def process_ingredient(ingredient: dict, aggregated: dict, exclude_pantry: bool):
    """Process a single ingredient and add to aggregated dict"""
    name = ingredient.get('nom', '').strip()
    quantity = float(ingredient.get('quantité', 0) or ingredient.get('quantite', 0))
    unit = ingredient.get('unité', '') or ingredient.get('unite', '')

    if not name:
        return

    # Skip pantry items if requested
    if exclude_pantry and is_pantry_item(name):
        return

    # Normalize quantities and units
    converted_qty, converted_unit = convert_quantity(quantity, unit)
    category = categorize_ingredient(name)

    # Create unique key for ingredient
    key = (name.lower(), converted_unit.lower())

    if key in aggregated:
        aggregated[key]['quantite'] += converted_qty
    else:
        aggregated[key] = {
            'nom': name,
            'quantite': converted_qty,
            'unite': converted_unit,
            'category': category
        }

def round_to_practical(quantity: float, unit: str) -> float:
    """Round quantities to practical amounts"""
    # For units (eggs, cans, etc.), round up to nearest whole number
    if unit in ['unité', 'unités', '', 'pièce', 'pièces']:
        return round(quantity + 0.49)  # Round 2.3 to 3, 2.7 to 3

    # For small quantities, keep 1 decimal
    if quantity < 10:
        return round(quantity, 1)

    # For larger quantities, round to nearest 5 or 10
    if quantity >= 100:
        return round(quantity / 10) * 10

    return round(quantity)
